torch/distributed/optim/named_optimizer.py

```diff
@@ -0,0 +1,244 @@
+            for group in param_groups:
+                for param in group["params"]:
+                    if param not in param_to_key:
+                        raise ValueError(f"Expect param name {param} found in param group but is missing.")
+                    param_keys_order.append(param_to_key[param])
+            self.param_keys_order = param_keys_order
+
+    def state_dict(self) -> Dict[str, Any]:
+        """
+        Return the state_dict of the optimzer. Instead of using number to index
+        parameters, we will use module fully qualifed name (FQN) as the key.
+        """
+        state_dict = self._optimizer.state_dict()
+        param_groups = state_dict["param_groups"]
+
+        ret_state = {
+            self.param_keys_order[st_key]: state_val
+            for st_key, state_val in state_dict["state"].items()
+        }
+
+        ret_groups = []
+        for group in param_groups:
+            param_keys = []
+            for param in group["params"]:
+                param_keys.append(self.param_keys_order[param])
+            ret_group = {"params": sorted(param_keys)}
+            for k, v in group.items():
+                if k != "params":
+                    ret_group[k] = deepcopy(v)
+            ret_groups.append(ret_group)
+
+        return {"state": ret_state, "param_groups": ret_groups}
+
+    def register_state_dict_pre(self):
+        raise NotImplementedError("register_state_dict_pre not supported yet and might be implemented soon.")
+
+    def register_state_dict_post(self):
+        raise NotImplementedError("register_state_dict_post not supported yet and might be implemented soon.")
+
+    def step(self):
+        """
+        Performs a single optimization step.
+
+        This will call :meth:`torch.optim.Optimizer.step` on the wrapped
+        optimizer.
+        """
+        self._optimizer.step()
+
+    def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
+        """
+        This public function defines the default behavior to load a state_dict
+        for named optimizer.
+
+        Sample Code
+        ```
+            my_model = MyModule()
+            optimizer = NamedOptimizer(my_model.named_parameters(), Adagrad)
+            ...
+
+            optim_state_dict = optimizer.state_dict()
+            ...
+
+            fs_storage_loader = torch.distributed.FileSystemLoader("/checkpoint/1")
+            torch.distributed.load_state_dict(
+                state_dict=optim_state_dict,
+                storage_reader=fs_storage_loader,
+            )
+
+            optimizer.load_state_dict(optim_state_dict)
+            ...
+        ```
+        Args:
+            state_dict (Dict[str, Any]) : A ``state_dict`` to load into the optimizer.
+                Note that this state dict update is performed in places.
+        """
+        # TODO: Need to handle the case when self._optimizer has not been initialized.
+        new_state_dict = self._optimizer.state_dict()
```

Comment:
Can we add a comment to rationale why getting the existing `state_dict` is required?